feat: add `_series_id` to tables on write

hiltontj opened the pull request:
Closes #24821

### Summary 
* New `_series_id` column is added to tables; this stores a 32 byte SHA256 hash of the tag set of a line of Line Protocol. The tag set is checked for sort order, then sorted if not already, before producing the hash.
* Unit tests were added to check hashing and sorting functions work.
* Tests that performed queries needed to be modified to account for the new `_series_id` column; in general, `SELECT *` queries were altered to use a select clause with specific column names.
* The Column limit was increased to 501 internally, to account for the new `_series_id` column, but the user-facing limit is still 500

### Other Considerations
* The `_series_id` column was added as a `String` column, and is stored using the HEX-encoded form of the hash; where as, it may be nice to use a column that can store the raw bytes. There is no suitable `ColumnType` variant in the [data_types crate](https://github.com/influxdata/influxdb3_core/blob/bb5eeb02614d32dbc244a17c7478030d5139d54b/data_types/src/columns.rs#L245-L253).
* When generating the `_series_id` hash, we [still need to check that it is sorted](https://github.com/influxdata/influxdb/pull/24842/files#diff-786737bb322e113951842b84574ed18ee04e82ec59e5acaa833f4e58acc716dbR754); which for very large tag sets could be inefficient.

mgattozzi commented on influxdb3_write/src/catalog.rs:
```diff
@@ -25,7 +25,7 @@ pub enum Error {
     #[error(
         "Update to schema would exceed number of tables limit of {} tables",
-        Catalog::NUM_TABLES_LIMIT
+        Catalog::NUM_TABLES_LIMIT - 1
```
You increased the limit for NUM_COLUMNS_PER_TABLE_LIMIT not the NUM_TABLES_LIMIT constant and so this change to subtract 1 would display the wrong number

hiltontj replied:
Oops, good catch, I will fix that.

mgattozzi commented on influxdb3_write/src/catalog.rs:
```diff
@@ -55,7 +56,10 @@ impl Catalog {
+    ///
+    /// The user-facing limit is 500, but we have it as 501 to account for the
+    /// generated _series_id column
+    pub(crate) const NUM_COLUMNS_PER_TABLE_LIMIT: usize = 501;
```
This change is good. Glad you caught that.

hiltontj replied:
I didn't catch it, the `limits` test did! 😄

mgattozzi commented on influxdb3_write/src/write_buffer/mod.rs:
```diff
@@ -507,21 +511,21 @@ pub(crate) fn validate_or_insert_schema_and_partitions(
     precision: Precision,
 ) -> Result<()> {
-    let table_name = line.series.measurement.as_str();
+    let table_name = line.series.measurement.to_string();
```
I don't think this needs to be a string. I don't see any changes you've made that would require that, unless something was causing the borrow checker to complain that I'm missing.

hiltontj replied:
I changed this to satisfy the borrower checker originally, but I will take a second look if this is needed.

mgattozzi commented on influxdb3_write/src/write_buffer/mod.rs:
```diff
@@ -727,6 +740,62 @@ pub(crate) struct TableBatchMap<'a> {
+    let bytes = if let Some(tag_set) = &mut series.tag_set {
+        if is_sorted(tag_set) {
+            // extract the raw string; if we are here, then the following `expect`s should
+            // never panic, there should be a tag set in the raw line protocol
```
Is this because we expect this to be a valid lp line?

hiltontj replied:
Yeah, at this point the `tag_set` is a `TagSet`, which is parsed from a valid line of LP, i.e., the `raw_line` that we pass forward here. So, if it exists, then we should be able to do the following extraction to generate the hash without allocating a new string.

mgattozzi commented on influxdb3_write/src/write_buffer/mod.rs:
```diff
@@ -727,6 +740,62 @@ pub(crate) struct TableBatchMap<'a> {
+        }
+    } else {
+        // for an empty tag set, we still need a series ID:
+        Sha256::digest("")
```
We should probably just impl Default for SeriesId, replace this with SeriesId::default(), and that method is just the precomputed hash of `Sha256::digest("")` to shave off some cpu cycles if a tag set isn't specified often.

hiltontj replied:
Nice, I like that, will give it a shot.

mgattozzi reviewed (changes requested):
Just a few changes needed, but overall 💯

pauldix reviewed (commented):
+1 on @mgattozzi's comments, but otherwise LGTM!

hiltontj commented:
@mgattozzi - I think I addressed your feedback in [`ba2e280`](https://github.com/influxdata/influxdb/pull/24842/commits/ba2e2803d666b05c2373c51ad4868986a7754099).

mgattozzi reviewed (approved):
LGTM
